### PR TITLE
[timeseries] typing improvements for gluonts models

### PR DESCRIPTION
**Pull request by canerturkmen**
*Issue #, if available:*

*Description of changes:*

Pyright typing improvements for abstract models and gluonts models

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py [hunk 1]
```diff
@@ -577,14 +607,23 @@ def _stack_sample_forecasts(self, forecasts: List[SampleForecast], item_ids: pd.
     def _stack_distribution_forecasts(self, forecasts: List[Forecast], item_ids: pd.Index) -> pd.DataFrame:
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:607]: Would it be possible to remove the need for the many asserts by:
- changing the type hint here to `forecasts: List["DistributionForecast"]` 
- adding a "masked" import of `DistributionForecast` on top of the file with
```python
if typing.TYPE_CHECKING:
    from gluonts.torch.model.forecast import DistributionForecast
```

Maybe this would also the need to add type hints for `item_id_to_forecast` and `dist_forecasts` below?

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py [hunk 2]
```diff
@@ -593,30 +632,30 @@ def stack_distributions(distributions: List[Distribution]) -> Distribution:
...
-            for key in dist.arg_constraints.keys():
+            for key in last_dist.arg_constraints.keys():
                 stacked_params[key] = torch.cat([p[key] for p in params_per_dist])
-            return dist.__class__(**stacked_params)
+            return last_dist.__class__(**stacked_params)
 
-        if not isinstance(forecasts[0].distribution, AffineTransformed):
-            raise AssertionError("Expected forecast.distribution to be an instance of AffineTransformed")
+        if not isinstance(dist_forecasts[0].distribution, AffineTransformed):
+            raise AssertionError()
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:640]: Could this be removed given the new assertion above? https://github.com/autogluon/autogluon/pull/4969/files#diff-28a0cc813c0c8a1cd39a4e8c358ee689718114207a2655af00cb4ae7675d7665R619-R621

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py [hunk 3]
```diff
@@ -629,9 +668,9 @@ def _gluonts_forecasts_to_data_frame(
 
         item_ids = forecast_index.unique(level=ITEMID)
         if isinstance(forecasts[0], SampleForecast):
-            forecast_df = self._stack_sample_forecasts(forecasts, item_ids)
+            forecast_df = self._stack_sample_forecasts(cast(List[SampleForecast], forecasts), item_ids)
         elif isinstance(forecasts[0], QuantileForecast):
-            forecast_df = self._stack_quantile_forecasts(forecasts, item_ids)
+            forecast_df = self._stack_quantile_forecasts(cast(List[QuantileForecast], forecasts), item_ids)
         elif isinstance(forecasts[0], DistributionForecast):
             forecast_df = self._stack_distribution_forecasts(forecasts, item_ids)
```
Comment (shchur) [inline timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:675]: What if instead of adding assertions for `all(isinstance(...))` in the respective methods, we use these checks as branches for the if conditions?
```python
if all(isinstance(f, SampleForecast) for f in forecasts):
   ....
else:
    raise AssertionError("Expected all forecasts to be of SampleForecast, QuantileForecast or DistributionForecast type")
```
If my intuitive understanding of how type checkers work is correct, this could save us some assertions / `cast` methods down the line.

**Review by shchur — APPROVED** [review]
LGTM! Just a few nits

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py (hunk 1 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:607]: cool, thanks!

File: timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py (hunk 3 above)
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/gluonts/abstract_gluonts.py:675]: I don't think it's that easy to pull off. For starters, the type checker won't be satisfied with the all(isinstance()) check. Hence skipping this for now.
